python/selfie-lib/selfie_lib/Snapshot.py

```diff
@@ -1,106 +1,69 @@
+from .ArrayMap import ArrayMap
 
 
 class Snapshot:
-    _subject: SnapshotValue
-    _facet_data: Dict[str, SnapshotValue]
-
     def __init__(
-        self, subject: SnapshotValue, facet_data: Dict[str, SnapshotValue]
+        self,
+        subject: SnapshotValue,
+        facet_data: Union[ArrayMap[str, SnapshotValue], None] = None,
     ) -> None:
         self._subject = subject
-        self._facet_data = facet_data
+        self._facet_data = facet_data if facet_data is not None else ArrayMap.empty()
 
     @property
-    def facets(self) -> OrderedDict:
-        return OrderedDict(sorted(self._facet_data.items()))
+    def facets(self) -> ArrayMap[str, SnapshotValue]:
+        return self._facet_data
 
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, Snapshot):
             return NotImplemented
         return self._subject == other._subject and self._facet_data == other._facet_data
 
     def __hash__(self) -> int:
-        return hash((self._subject, frozenset(self._facet_data.items())))
+        return hash((self._subject, tuple(self._facet_data.items())))
 
     def plus_facet(
         self, key: str, value: Union[bytes, str, SnapshotValue]
     ) -> "Snapshot":
-        """
-        Add a facet with the given key and value to the snapshot.
-        The value can be bytes, a string, or a SnapshotValue.
-        """
-        # Ensure value is converted to SnapshotValue if it's not already one
         if isinstance(value, bytes):
             value = SnapshotValueBinary(value)
         elif isinstance(value, str):
             value = SnapshotValueString(value)
         elif not isinstance(value, SnapshotValue):
             raise TypeError("Value must be either bytes, str, or SnapshotValue")
-
-        # Now we can safely pass a SnapshotValue instance to _plus_facet
-        return self._plus_facet(key, value)
-
-    def _plus_facet(self, key: str, value: SnapshotValue) -> "Snapshot":
-        """
-        The actual implementation to add a facet. This method should only be called
-        with a SnapshotValue instance as value.
-        """
-        if not key:
-            raise ValueError("The empty string is reserved for the subject.")
-        facet_data = dict(self._facet_data)
-        facet_data[key] = value  # Here, value is guaranteed to be a SnapshotValue
-        return Snapshot(self._subject, facet_data)
+        new_facet_data = self._facet_data.plus(key, value)
+        return Snapshot(self._subject, new_facet_data)
 
     def plus_or_replace(self, key: str, value: SnapshotValue) -> "Snapshot":
         if not key:
             return Snapshot(value, self._facet_data)
-        facet_data: Dict[str, SnapshotValue] = dict(self._facet_data)
-        facet_data[self._unix_newlines(key)] = value
-        return Snapshot(self._subject, facet_data)
+        return Snapshot(self._subject, self._facet_data.plus(key, value))
```

Comment:
`ArrayMap.plus` throws an exception if the key is already present. Here is the reference implementation

https://github.com/diffplug/selfie/blob/b15b9126c981bb5b0c2c62c3fdfecbacfa64045a/jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/SnapshotFile.kt#L80-L82

Looks like our `ArrayMap` needs `plusOrNoOpOrReplace`. That would be a good thing to include in this PR.